hazelcast/src/main/java/com/hazelcast/jet/impl/AbstractJobProxy.java

```diff
@@ -294,6 +301,16 @@ private void terminate(TerminationMode mode) {
         }
     }
 
+    @Override
+    public UUID addStatusListener(@Nonnull JobStatusListener listener) {
+        try {
+            return doAddStatusListener(listener);
+        } catch (JobNotFoundException ignored) {
+            throw JobEventService.illegalStateException(
+                    future.isCompletedExceptionally() ? FAILED : COMPLETED);
```

Comment:
One can only obtain a proxy for an existing job, so `JobNotFoundException` can only occur if the job is completed and removed by `JobCoordinationService`. At least, this is the designed behavior and it is immune to node failures since `JobRepository` stores job records, execution records and results in IMap's.

Reply:
are you sure that there is no race condition possible in `addStatusListener` between `JobNotFoundException` and `future` completion? Even if JobCoordinationService uses locks, there are still many job coordinator threads. Is it possible that client gets addStatusListener response before job completion/failure message? Are those messages (addStatusListener response and job join response) guaranteed to be handled in order?